build_tools/release_automations/create_release_branch.py

```diff
@@ -0,0 +1,510 @@
+#!/usr/bin/env python3
+"""
+This script automates the creation of release branches for theRock projects and their submodules.
+It supports branching from a specified source branch, including or excluding specific projects,
+and optionally notifying stakeholders via email.
+
+Main Features:
+- Clones or uses an existing source tree for theRock and its submodules.
+- Checks out the specified source branch for all relevant projects.
+- Determines projects to branch based on .gitmodules and user input.
+- Creates a plan for branching, including checks for branch existence and push permissions.
+- Displays a summary of planned actions
+- Executes branch creation and pushes to remote repositories.
+- Sends notifications to specified email addresses.
+
+Usage:
+    python create_release_branch.py --release_branch <branch> [options]
+
+Options:
+    --release_branch         Name of the branch to create (required).
+    --apitoken              API token for authentication with upstream repositories.
+    --dry-run                Executed all part of the code except git push
+
+Note:
+    - This script assumes a specific repository structure and the presence of .gitmodules.
+"""
+import argparse
+import configparser
+import logging
+import shutil
+import re
+import sys
+import subprocess
+import tempfile
+from pathlib import Path
+import shlex
+from typing import Dict, Tuple, Any
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
+
+
+class RockBranchingAutomation:
+    """Class for rock branching automation"""
+
+    def __init__(self, cli_args):
+        self.release_branch = cli_args.release_branch
+        self.apitoken = cli_args.apitoken
+        self.dry_run = getattr(cli_args, "dry_run", False)
+
+    def log(self, *args, **kwargs):
+        """
+        log function to print and flush output
+        """
+        print(*args, **kwargs)
+        sys.stdout.flush()
+
+    def exec(self, args: list[str | Path], cwd: Path):
+        """
+        Executes a command in a subprocess.
+        """
+        cmd = [str(arg) for arg in args]
+        self.log(f"++ Exec [{cwd}]$ {shlex.join(cmd)}")
+        sys.stdout.flush()
+        subprocess.check_call(args, cwd=str(cwd), stdin=subprocess.DEVNULL)
+
+    def checkout_source(self) -> Tuple[Path, Dict[str, str]]:
+        """
+        Clone theRock repository and return its path along with submodules dict.
+        """
+        the_rock_url = "https://github.com/ROCm/TheRock.git"
+        the_rock_dir_path = Path.cwd() / "rock-branching"
```

Comment:
if this is aprt of the ci you do not need a special `rock-branching` directory. just take the default of `git clone`